File: leet-code-rs/src/solutions/349_两个数组的交集.rs
/*
 * @lc app=leetcode.cn id=349 lang=rust
 *
 * [349] 两个数组的交集
 */

// @lc code=start
use std::collections::HashSet;
impl Solution {
    pub fn intersection(nums1: Vec<i32>, nums2: Vec<i32>) -> Vec<i32> {
        // nums1
        //     .into_iter()
        //     .collect::<HashSet<i32>>()
        //     .intersection(&nums2.into_iter().collect::<HashSet<i32>>())
        //     .map(|&x| x)
        //     .collect()
        let mut st = nums1.into_iter().collect::<HashSet<_>>();
        let mut ans = vec![];

        for x in nums2 {
            if st.remove(&x) {
                // x 在 st 中
                ans.push(x);
            }
        }
        ans
    }
}
// @lc code=end

#[allow(unused)]
struct Solution;

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_intersection() {
        //assert_eq!(Solution::intersection(vec![]),[]);
        assert!(true)
    }
}

File: leet-code/349.两个数组的交集.js
/*
 * @lc app=leetcode.cn id=349 lang=javascript
 *
 * [349] 两个数组的交集
 */

// @lc code=start
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var intersection = function(nums1, nums2) {
  const nums1Set = new Set(nums1);
  const resSet = new Set();
  // for(const n of nums2) {
  //     nums1Set.has(n) && resSet.add(n);
  // }
  // 循环 比 迭代器快
  for(let i = nums2.length - 1; i >= 0; i--) {
      nums1Set.has(nums2[i]) && resSet.add(nums2[i]);
  }
  return Array.from(resSet);
};
// @lc code=end

